pkg/controller/kubelet-config/kubelet_config_bootstrap.go

```diff
@@ -0,0 +1,175 @@
+			if err != nil {
+				return nil, err
+			}
+			if kubeletConfig.Spec.TLSSecurityProfile != nil {
+				// Inject TLS Options from Spec
+				observedMinTLSVersion, observedCipherSuites := getSecurityProfileCiphers(kubeletConfig.Spec.TLSSecurityProfile)
+				originalKubeConfig.TLSMinVersion = observedMinTLSVersion
+				originalKubeConfig.TLSCipherSuites = observedCipherSuites
+			}
+
+			kubeletIgnition, logLevelIgnition, autoSizingReservedIgnition, err := generateKubeletIgnFiles(kubeletConfig, originalKubeConfig, userDefinedSystemReserved)
+			if err != nil {
+				return nil, err
+			}
+
+			tempIgnConfig := ctrlcommon.NewIgnConfig()
+			if autoSizingReservedIgnition != nil {
+				tempIgnConfig.Storage.Files = append(tempIgnConfig.Storage.Files, *autoSizingReservedIgnition)
+			}
+			if logLevelIgnition != nil {
+				tempIgnConfig.Storage.Files = append(tempIgnConfig.Storage.Files, *logLevelIgnition)
+			}
+			if kubeletIgnition != nil {
+				tempIgnConfig.Storage.Files = append(tempIgnConfig.Storage.Files, *kubeletIgnition)
+			}
+
+			rawIgn, err := json.Marshal(tempIgnConfig)
+			if err != nil {
+				return nil, err
+			}
+			managedKey, err := generateSuffixManagedKeyKubelet(pool, kubeletConfig, generatedManagedKeys)
+			if err != nil {
+				return nil, err
+			}
+			var (
+				mc *mcfgv1.MachineConfig
+				ok bool
+			)
+			mc, ok = generatedMCs[managedKey]
+			if !ok {
+				mc, err = ctrlcommon.MachineConfigFromIgnConfig(role, managedKey, kubeletIgnition)
+				if err != nil {
+					return nil, err
+				}
+				generatedMCs[managedKey] = mc
+			}
+
+			mc.Spec.Config.Raw = rawIgn
+			mc.SetAnnotations(map[string]string{
+				ctrlcommon.GeneratedByControllerVersionAnnotationKey: version.Hash,
+			})
+			oref := metav1.OwnerReference{
+				APIVersion: controllerKind.GroupVersion().String(),
+				Kind:       controllerKind.Kind,
+			}
+			mc.SetOwnerReferences([]metav1.OwnerReference{oref})
+			res = append(res, mc)
+		}
+	}
+	return res, nil
+}
+
+// generateSuffixManagedKeyKubelet generates the number suffixed machine config name for a CR during bootstrap
+func generateSuffixManagedKeyKubelet(pool *mcfgv1.MachineConfigPool, cfg *mcfgv1.KubeletConfig, managedKeys map[kubeletconfigKey]string) (string, error) {
+	// If there is no kubelet config in the list, return the default MC name with no suffix
+	var cfgCnt int
+	for k := range managedKeys {
+		if k.poolName == pool.Name {
+			cfgCnt++
+		}
+	}
+	if cfgCnt == 0 {
+		managedKey, err := ctrlcommon.GetManagedKey(pool, nil, "99", "kubelet", getManagedKubeletConfigKeyDeprecated(pool))
+		if err != nil {
+			return "", err
+		}
+		managedKeys[kubeletconfigKey{pool.Name, cfg.Name}] = managedKey
+		return managedKey, nil
+	}
+	// If we find a matching kubelet config name and return that existed managedKey as this kubelet config exists
+	// we are probably doing an update action on it
+	if managedKey, ok := managedKeys[kubeletconfigKey{pool.Name, cfg.Name}]; ok {
```

Comment:
So this will only happen if you supply multiple kubeletconfigs with the same name during bootstrap, correct? And then we are just taking on the later one based on... whichever file was parsed later (I assume, since there is no sorting)

Technically this matches normal in-cluster behaviour when you overwrite an existing KC, but I wonder if we should just error here instead since we can be a little more restrictive during bootstrap.

Reply:
> So this will only happen if you supply multiple kubeletconfigs with the same name during bootstrap, correct? And then we are just taking on the later one based on... whichever file was parsed later (I assume, since there is no sorting)
> 

Yes, there's no sorting to decide which one is the latter one that will be applied.

> Technically this matches normal in-cluster behaviour when you overwrite an existing KC, but I wonder if we should just error here instead since we can be a little more restrictive during bootstrap.

I agree. Fixed to return an error.